### PR TITLE
ARROW-6157: [C++] Array data validation

### PR DESCRIPTION
**Pull request by pitrou**
Add a method ValidateFull() on arrays, batches etc. which does O(N) data validation
for a few types (list, union, dictionary).

Also, fix the assumptions about union arrays to match official semantics.

### CODE REVIEW

File: cpp/src/arrow/array.h [hunk 1]
```diff
@@ -1165,12 +1166,21 @@ class ARROW_EXPORT UnionArray : public Array {
   /// Note that this buffer does not account for any slice offset
   std::shared_ptr<Buffer> type_ids() const { return data_->buffers[1]; }
 
+  const type_id_t* raw_type_ids() const { return raw_type_ids_ + data_->offset; }
```
Comment (pitrou) [inline cpp/src/arrow/array.h]: @wesm I was wondering... should we rename these methods to `type_codes` and `raw_type_codes`? There is a bit of a mixed terminology here (`UnionType` uses `type_codes`, also `type_id` is used for something different).

**Comment by pitrou** [comment]
cc @wesm For the union changes.

File: cpp/src/arrow/array.h (hunk 1 above)
Comment (wesm) [reply cpp/src/arrow/array.h]: I'd be in favor of changing the names to be more clear and where relevant, conforming to the specification

File: cpp/src/arrow/type.cc [hunk 2]
```diff
@@ -288,13 +288,25 @@ std::string DurationType::ToString() const {
...
+constexpr int UnionType::kInvalidChildId;
+
 UnionType::UnionType(const std::vector<std::shared_ptr<Field>>& fields,
-                     const std::vector<uint8_t>& type_codes, UnionMode::type mode)
-    : NestedType(Type::UNION), mode_(mode), type_codes_(type_codes) {
+                     const std::vector<int8_t>& type_codes, UnionMode::type mode)
+    : NestedType(Type::UNION),
+      mode_(mode),
+      type_codes_(type_codes),
+      child_ids_(kMaxTypeId + 1, kInvalidChildId) {
```
Comment (wesm) [inline cpp/src/arrow/type.cc]: Do I have it right that this is adding `sizeof(int) * (kMaxTypeId + 1)` (so 512 or 1024 bytes depending on the system) to the footprint of every UnionType instance?

**Review by wesm — COMMENTED** [review]
Thanks for doing all this, definitely of high value to provide deeper validation than we had previously.

I left a couple comments about things that stood out but everything else looked in line with expectations

File: cpp/src/arrow/array.h (hunk 1 above)
Comment (pitrou) [reply cpp/src/arrow/array.h]: Well, the specification uses "types" and "type ids" :-/

File: cpp/src/arrow/array/validate.cc [hunk 3]
```diff
@@ -0,0 +1,440 @@
...
+  template <typename ArrayType>
+  Status ValidateOffsets(const ArrayType& array) {
+    auto prev_offset = array.value_offset(0);
+    for (int64_t i = 1; i <= array.length(); ++i) {
+      auto current_offset = array.value_offset(i);
+      if (array.IsNull(i - 1) && current_offset != prev_offset) {
+        return Status::Invalid("Offset invariant failure at: ", i,
+                               " inconsistent value_offsets for null slot",
+                               current_offset, "!=", prev_offset);
+      }
```
[the remark this answers is not in the document]
Comment (pitrou) [reply cpp/src/arrow/array/validate.cc]: This code is simply moved around from `array.cc`.
The spec does not spell it out explicitly but the examples are compliant with this expectation:
https://arrow.apache.org/docs/format/Columnar.html#variable-size-list-layout

File: cpp/src/arrow/type.cc (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/type.cc]: Yes. I think this is simpler than having to lazily-initialize it.

**Comment by jorisvandenbossche** [comment]
Should the python constructors do a cheap or a full validation? (since `full=False`is the default, I assume they will now do a cheap validation) 

So that means that the initial example in the issue (https://issues.apache.org/jira/browse/ARROW-6157) with `pa.UnionArray.from_dense` can still generate a invalid array that will segfault on certain operations? 
That might be fine (it should probably at least be possible to not do the full validation, in case you know you have valid array data), but maybe we should then update some of those docstrings to note that?

**Comment by pitrou** [comment]
Full validation is essentially O(n), so it sounds undesirable to do it by default. I don't know what @wesm thinks.

**Comment by pitrou** [comment]
Rebased. Will merge when green.

**Pull request closed without merging**